src/main/scala/is/hail/expr/ir/Simplify.scala

```diff
@@ -140,6 +141,16 @@ object Simplify {
         val makeStructFields = fields.toMap
         MakeStruct(fieldNames.map(f => f -> makeStructFields(f)))
 
+      case x@SelectFields(InsertFields(struct, insertFields), selectFields) =>
+        val selectSet = selectFields.toSet
+        val insertFields2 = insertFields.filter { case (fName, _) => selectSet.contains(fName) }
+        val insertSet = insertFields2.map(_._1).toSet
+        val structSet = struct.typ.asInstanceOf[TStruct].fieldNames.toSet
+        val selectFields2 = selectFields.filter(s => !insertSet.contains(s) || (insertSet.contains(s) && structSet.contains(s)))
```

Comment:
I think this is identical to just `structSet.contains(s)`?

Reply:
I think the entire condition is equivalent---`val selectFields2 = selectFields.filter(structSet.contains(_))` and you could get rid of `insertSet`---since the original SelectFields is invalid if it's neither a field in the original struct or a field that you've inserted. But this is also fine, and we don't have to rely on the correctness of the SelectFields type.